### DISCUSSION TITLE
Arg mapping to complex arg values results in cyclic object value TypeError

### DISCUSSION THREAD

**Discussion by nihiluis** [storybookjs/storybook#25463] [Help] [answered]
### Summary

Hi, I'm trying to add a component to Storybook which takes a "complex" control arg, meaning it can't be serialized. Based on the docs (https://storybook.js.org/docs/writing-stories/args#mapping-to-complex-arg-values) and my search you are supposed to use the mapping feature, but it results in a `TypeError: cyclic object value`. Perhaps I have the same issue as here: https://github.com/storybookjs/storybook/discussions/19199

Is that code correct here?

```ts
export default {
  title: 'PrimeNg/InputNgComponent',
  component: InputNgComponent,
  argTypes: {
    control: {
      options: ['InputTextNoErrors'],
      mapping: {
        InputTextNoErrors: inputTextControl
      },
      control: 'select'
    }
  }
} as Meta;

export const Primary: StoryObj<InputNgComponent> = {
  args: {
    label: 'Button',
    control: "InputTextNoErrors"
  }
};
```

### Additional information

Storybook version 7.6.7

**Comment by Marklb** [comment]
Based on the snippet you provided, I think your story looks fine, but the one in the referenced discussion looks like it may be defined incorrectly. Can you provide a repro that gives the error?

I attempted to create a Stackblitz repro, based on what I assume you are trying to do from what you provided, but I don't get an error. https://stackblitz.com/edit/github-ypaznh?file=src%2Fapp%2Fexample%2Ftest-form.stories.ts,src%2Fapp%2Fexample%2Finput-ng.component.ts&preset=node

**Comment by nihiluis** [comment]
Thank you for the pointer and repro. I'm hardly familiar with the `control` arg that is used, I assume my control object has a circular reference in it which the one in your repro doesn't contain. Is that a nono for Storybook?

**Reply by Marklb** [reply, marked as answer]
I didn't think about FormControl only having a circular reference if it has a parent. Now, I do get the error. I thought that was a bug that had been fixed, but I guess not and I can't seem to find the issue about it. If I find the issue then I will update this comment.

The error is happening, when Storybook tries to convert the inputs to a readable value in the source that is displayed, when clicking "Show Code" on a docs page.

The workaround is to provide yours own template. `argsToTemplate` should generated the inputs/outputs, with the props variable that is going to actually be used, instead of a stringified readable value. So, the following should generate a template very close to what Storybook is actually using as the story's component template, when it generates the template for you.

```ts
export const Primary: StoryObj<StoryComponent> = {
  render: args => ({
    props: args,
    template: `<ex-input-ng ${argsToTemplate(args)}></ex-input-ng>`,
  }),
  args: {
    label: 'Button',
    control: 'InputTextNoErrors',
  },
};
```
Here is my updated Stackblitz: https://stackblitz.com/edit/github-ypaznh-tcotwg?file=src%2Fapp%2Fexample%2Ftest-form.stories.ts,src%2Fapp%2Fexample%2Finput-ng.component.ts&preset=node

**Reply by nihiluis** [reply]
Thank you! Works well.
